qiita_ware/commands.py

```diff
@@ -21,6 +25,144 @@
+    # if port is '' Python 2.7.6 will raise an error
+    try:
+        port = p_url.port
+    except Exception:
+        port = 22
+    username = p_url.username
+
+    if scheme == 'scp' or scheme == 'sftp':
+
+        # if port not specified, use default 22 as port
+        if port is None:
+            port = 22
+
+        # step 1: both schemes require an SSH connection
+        ssh = SSHClient()
+        ssh.set_missing_host_key_policy(AutoAddPolicy)
+
+        # step 2: connect to fileserver
+        key = RSAKey.from_private_key_file(private_key)
+        ssh.connect(hostname, port=port, username=username,
+                    pkey=key, look_for_keys=False)
+        return ssh
+    else:
+        raise ValueError(
+            'Not valid scheme. Valid options are ssh and scp.')
+
+
+def _list_valid_files(ssh, directory):
+    """Gets a list of valid study files from ssh session
+
+    Parameters
+    ----------
+    ssh : paramiko.SSHClient
+        An initializeed ssh session
+    directory : str
+        the directory to search for files
+
+    Returns
+    -------
+    list of str
+        list of valid study files (basenames)
+    """
+
+    valid_file_extensions = tuple(qiita_config.valid_upload_extension)
+    sftp = ssh.open_sftp()
+    files = sftp.listdir(directory)
+
+    valid_files = [f for f in files if f.endswith(valid_file_extensions)]
+    sftp.close()
+    return valid_files
+
+
+def list_remote(URL, private_key):
+    """Retrieve valid study files from a remote directory
+
+    Parameters
+    ----------
+    URL : str
+        The url to the remote directory
+    private_key : str
+        Path to the private key used to authenticate connection
+
+    Returns
+    -------
+    list of str
+        list of files that are valid study files
+
+    Notes
+    -----
+    Only the allowed extensions described by the config file
+    will be listed.
+    """
+    p_url = urlparse(URL)
+    directory = p_url.path
+    ssh = _ssh_session(p_url, private_key)
+    valid_files = _list_valid_files(ssh, directory)
+    ssh.close()
+    return valid_files
+
+
+def download_remote(URL, private_key, destination):
+    """Add study files by specifying a remote directory to download from
+
+    Parameters
+    ----------
+    URL : str
+        The url to the remote directory
+    private_key : str
+        Path to the private key used to authenticate connection
+    destination : str
+        The path to the study upload folder
+    """
+
+    # step 1: initialize connection and list valid files
+    p_url = urlparse(URL)
+    ssh = _ssh_session(p_url, private_key)
+
+    directory = p_url.path
+    valid_files = _list_valid_files(ssh, directory)
+    file_paths = [join(directory, f) for f in valid_files]
+
+    # step 2: download files
+    scheme = p_url.scheme
+    # note that scp/sftp's code seems similar but the local_path/localpath
+    # variable is different within the for loop
+    if scheme == 'scp':
```

Comment:
*From my previous review*

I think this can be broken down to be a bit simpler, for example:

```python
     if scheme == 'scp':
        client = SCPClient(ssh.get_transport())
     elif scheme == 'sftp':
        client = ssh.open_sftp()

     for f in file_paths:
         download = partial(client.get,
                            local_path=join(destination, basename(f)))
         download(f)
```

Also it should throw an exception if the scheme is not supported.

Reply:
Sorry, forgot to mention this.
1. The split is not really possible because, as the note above says, scp and sftp use a different variable local_path vs localpath; and couldn't figure out how to make it play nice with your above come - note that most of the pushes were due errors on these lines.
2. The exception is gonna be thrown a few lines above, when: `ssh = _ssh_session(p_url, private_key)`

Reply:
I see, thanks for pointing that out. Subtle differences in the api 😑